distuv/gamma.go

```diff
@@ -95,56 +95,112 @@ func (g Gamma) Rand() float64 {
-			t1 = math.Exp(t / a)
-			if 2*math.Log(u) <= t-t1 {
-				break
+			e := exprnd()
+			var z float64
+			if e >= -lr {
+				z = e + lr
+			} else {
+				z = -exprnd() / lambda
+			}
+			lh := lc - z - math.Exp(-z/a)
+			var lEta float64
+			if z >= 0 {
+				lEta = lc - z
+			} else {
+				lEta = lc + lw + math.Log(lambda) + lambda*z
+			}
+			if lh-lEta > -exprnd() {
+				return math.Exp(-z/a) / b
 			}
 		}
-		if a >= 0.01 {
-			return t1 / b
+	case a >= 0.3 && a < 1:
+		// Generate using:
+		//  Kundu, Debasis, and Rameshwar D. Gupta. "A convenient way of generating
+		//  gamma random variables using generalized exponential distribution."
+		//  Computational Statistics & Data Analysis 51.6 (2007): 2796-2802.
+
+		// TODO(btracey): Change to using Algorithm 3 if we can find the bug in
+		// the implementation below.
+
+		// Algorithm 2.
+		alpha := g.Alpha
+		a := math.Pow(1-math.Exp(-1.0/2), alpha) / (math.Pow(1-math.Exp(-1.0/2), alpha) + alpha*math.Exp(-1)/math.Pow(2, alpha))
+		b := math.Pow(1-math.Exp(-1.0/2), alpha) + alpha*math.Exp(-1)/math.Pow(2, alpha)
+		var x float64
+		for {
+			u := unifrnd()
+			if u <= a {
+				x = -2 * math.Log(1-math.Pow(u*b, 1/alpha))
+			} else {
+				x = -math.Log(math.Pow(2, alpha) / alpha * b * (1 - u))
+			}
+			v := unifrnd()
+			if x <= 1 {
+				if v <= math.Pow(x, alpha-1)*math.Exp(-x/2)/(math.Pow(2, alpha-1)*math.Pow(1-math.Exp(-x/2), alpha-1)) {
+					break
+				}
+			} else {
+				if v <= math.Pow(x, alpha-1) {
+					break
+				}
+			}
 		}
-		return t / a / b
+		return x / g.Beta
+
+		/*
+			//  Algorithm 3.
+			d := 1.0334 - 0.0766*math.Exp(2.2942*alpha)
+			a := math.Pow(2, alpha) * math.Pow(1-math.Exp(-d/2), alpha)
+			b := alpha * math.Pow(d, alpha-1) * math.Exp(-d)
+			c := a + b
+			var x float64
+			for {
+				u := unifrnd()
+				if u <= a/(a+b) {
+					x = -2 * math.Log(1-math.Pow(c*u, 1/a)/2)
+				} else {
+					x = -math.Log(c * (1 - u) / (alpha * math.Pow(d, alpha-1)))
+				}
+				v := unifrnd()
+				if x <= d {
+					if v <= (math.Pow(x, alpha-1)*math.Exp(-x/2))/(math.Pow(2, alpha-1)*math.Pow(1-math.Exp(-x/2), alpha-1)) {
+						break
+					}
+				} else {
+					if v <= math.Pow(d/x, 1-alpha) {
+						break
+					}
+				}
+			}
+			return x / g.Beta
+		*/
 	case a > 1:
-		// Generate using
-		//  Martino, Luca, and David Luengo. "Extremely efficient generation of
-		//  Gamma random variables for α >= 1." arXiv preprint arXiv:1304.3800 (2013).
-		ap := math.Floor(a)
-		var bp, lkp float64
-		// The paper says ap < 2, but ap must be an integer at least 1.
-		if ap == 1 {
-			bp = b / a
-			lkp = (1 - a) + (a-1)*(a/b)
-		} else {
-			bp = b * (ap - 1) / (a - 1)
-			lkp = (ap - a) + (a-ap)*(a-1)/b
-		}
+		// Generate using:
+		//  Marsaglia, George, and Wai Wan Tsang. "A simple method for generating
+		//  gamma variables." ACM Transactions on Mathematical Software (TOMS)
+		//  26.3 (2000): 363-372.
+		d := a - 1.0/3
+		sqrt9d := 3 * math.Sqrt(d)
```

Comment:
Make this the reciprocal since `a*b` is faster than `a/b`. `c := 1 / (3*math.Sqrt(d))` and use below. This also matches the paper better.